adr: make monitoring forwarding storage independent (#56)
elastisys/compliantkubernetes: docs/adr/0007-make-monitoring-forwarders-storage-independent.md

* Buffered monitoring information is lost if node is lost.
* emptyDir can cause disk pressure. This can be handled by alerting on low disk space.

## Pros and Cons of the Options

### Use underlying storage provider

* Good, because the forwarder can be restarted on any node.
* Good, because the buffer can be large.
* Good, because no buffered monitoring information is lost if a node goes down.
* Good, because buffered monitoring information is preserved if the forwarder is redeployed.
* Bad, because non-node-local storage is generally slower. Note, however, that at least SafeSpring and CityCloud use a central Ceph storage cluster for the VM's boot disk, which wipes out node-local's storage advantage.)
* Bad, because the forwarder will fail if storage provider goes down. This is especially problematic for Exoscale, bare-metal and BYO-VMs.
* Bad, because the forwarder cannot monitor the storage provider (circular dependency).
* Bad, because setting right ownership requires init containers or [alpha features](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods).

### Use Local Persistent Volumes

> * Bad, because the forwarder cannot be restarted on any node without manual action: "[i]f a node becomes unhealthy, then the local volume becomes inaccessible by the pod. The pod using this volume is unable to run.".

Reviewer: ```suggestion
* Bad, because the forwarder cannot be restarted on any node without manual action: "if a node becomes unhealthy, then the local volume becomes inaccessible by the pod. The pod using this volume is unable to run.".
```

* Good, because the forwarder can monitor the storage provider (no circular dependency).
* Bad, because local persistent storage requires an additional configuration step.
* Bad, because setting right ownership requires init containers or [alpha features](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods).

### Use emptyDir

* Good, because the forwarder can be restarted on any node without manual action.
* Bad, because the amount of forwarding depends on the node's local disk size.
* Bad, because buffered monitoring information is lost if the forwarder's node goes down.
* Bad, because buffered monitoring information is lost if the forwarder is (not carefully enough) redeployed.
* Good, because node-local storage is generally faster.
* Good, because the forwarder will survive failure of storage provider.
* Good, because the forwarder can monitor the storage provider (no circular dependency).
* Good, because works out of the box.
* Good, because it integrates nicely with `securityContext`.

### Use hostPath

Similar to Local Persistent Volumes, but

> * Worse, because if the forwarder is redeployed on a new node, buffering information may appear/disappear.

Reviewer: Isn't this the same for local persistent volumes? AFAIK the data there is also tied directly to a node.
I'd say host path is only worse since it does not force you to use PVCs and thereby it might not be as controlled.

Author: My understanding is that `hostPath` really only mounts the path that is on the current node, without the scheduler caring. In contrast, local persistent volumes binds the Pod to the Node:

> With HostPath volumes, a pod referencing a HostPath volume may be moved by the scheduler to a different node resulting in data loss. But with Local Persistent Volumes, the Kubernetes scheduler ensures that a pod using a Local Persistent Volume is always scheduled to the same node.

https://kubernetes.io/blog/2019/04/04/kubernetes-1.14-local-persistent-volumes-ga/
